ext/stackprof/stackprof.c

```diff
@@ -411,26 +421,47 @@ stackprof_record_sample()
     }
 }
 
+void
+stackprof_record_sample()
+{
+    int num = rb_profile_frames(0, sizeof(_stackprof.frames_buffer) / sizeof(VALUE), _stackprof.frames_buffer, _stackprof.lines_buffer);
+    stackprof_record_sample_for_stack(num);
+}
+
+void
+stackprof_record_gc_sample()
+{
+    _stackprof.frames_buffer[0] = _stackprof.fake_gc_frame;
+    _stackprof.lines_buffer[0] = 0;
+    stackprof_record_sample_for_stack(1);
+}
+
 static void
-stackprof_job_handler(void *data)
+stackprof_job_handler(void *is_gc)
 {
     static int in_signal_handler = 0;
     if (in_signal_handler) return;
     if (!_stackprof.running) return;
 
     in_signal_handler++;
-    stackprof_record_sample();
+    if ((int)is_gc) {
+	stackprof_record_gc_sample();
+    } else {
+	stackprof_record_sample();
+    }
     in_signal_handler--;
 }
 
 static void
 stackprof_signal_handler(int sig, siginfo_t *sinfo, void *ucontext)
 {
     _stackprof.overall_signals++;
-    if (rb_during_gc())
-	_stackprof.during_gc++, _stackprof.overall_samples++;
-    else
-	rb_postponed_job_register_one(0, stackprof_job_handler, 0);
+    if (rb_during_gc()) {
+	_stackprof.during_gc++;
+	rb_postponed_job_register_one(0, stackprof_job_handler, (void*)1);
```

Comment:
Good catch. The list of signal-safe functions is very small, so calling into rb_hash_aset etc from the signal handler was not going to work reliably.